src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs

```diff
@@ -132,7 +132,7 @@ public override bool TryReadInternal(out ReadResult readResult)
                 KestrelBadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTimeout);
             }
 
-            TryStart();
+            TryStartAsync();
```

Comment:
Isn't this a behavior change to go from blocking to fire-and-forget?  Can TryStartAsync throw asynchronously like if the client has disconnected?

Reply:
If `Write100ContinueAsync()` could realistically return an incomplete ValueTask this is a change in behavior. Not sure how that could ever happen though. If it could happen, not blocking is waaay better than block because at least you return the thread to the pool. TryStartAsync should never be able to throw asynchronously because it's a write without a canceled token.

Reply:
Add an assert. I dislike the fire and forget as well unless we have an explicit check that it's always completed.

Reply:
https://github.com/dotnet/aspnetcore/pull/31650/commits/70e8ee1a5fe32cbb628bd809e0f74c55fefdf6f3

Reply:
Whoops, plus https://github.com/dotnet/aspnetcore/pull/31650/commits/a10c9d999e6d687d85c3d057f6fee35e8fef03c1

Reply:
I love that bug 😄

Reply:
> I dislike the fire and forget

I totally agree. I also dislike fire-and-forget with a passion, but what's even worse is blocking on I/O. This isn't the [only](https://github.com/dotnet/aspnetcore/blob/1d7443d1823a63b60f94ec9e93b1c4ca8e635aae/src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/StreamInputFlowControl.cs#L60) [place](https://github.com/stephentoub/aspnetcore/blob/b5145963a8a44832bf2d49a0beb1e6de0acb4cd5/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs#L147-L148) we don't await  output-writing `ValueTask`s we expect to be usually completed.

Now unlike with HTTP/2 where you could conceive of a scenario where output backpressure has built up prior to sending a `WINDOW_UPDATE` or `RST_STREAM`, it's hard to come up with one for an HTTP/1.1 "100 continue" response. That said, it's not a mathematical invariant. I could write a unit test with a custom transport and write buffering disabled to force it. For that reason, we should have this Debug.Assert like we don't for `WINDOW_UPDATE` or `RST_STREAM`. If we think it's worthwhile we can store the task and await it in ConsumeAsync(), but I don't think that's necessary.

Reply:
That makes sense to me - if it matches what we do for `RST_STREAM` and `WINDOW_UPDATE` seems like we can remove the assert here